blender_imgui.py

```diff
@@ -338,18 +344,66 @@ def handler_remove(self, handle):
         if not self.callbacks:
             self.shutdown_imgui()
 
+    lastLetThroughTime = 0
+    def returnTrueXAmountOfTimesPerSecondToCreateSolidFramerate(self, framerate: int, frameNumber: list[int] = [0]):
+        """frame number should be something like: frameNumber = [0]. 
+        It will set index 0 of the list to a number that will always be incremented by 1 + any number of missed frames since last the function was called.
+
+        Example usecase of this function would be:
+
+        while(True):
+        frameNumber = [0]
+        if returnTrueXAmountOfTimesPerSecondToCreateSolidFramerate(4, frameNumber):
+            print(frameNumber)
+
+        which would print the current frame number 4 times every second. Each frame number would always be 1 greater than the previous."""
+        import time
+        import math
+        # framerate = 4
+        def formula():
+            return math.floor(time.time()*framerate)
+
+        currentTime = formula()
+        # try:
+        self.lastLetThroughTime
+        # except:
+        #     lastLetThroughTime = 0
+        # print(lastLetThroughTime)
+        if self.lastLetThroughTime != currentTime:
+            self.lastLetThroughTime = currentTime
+            frameNumber[0] = currentTime
+            return True
+        frameNumber[0] = currentTime
+        return False
+
     def draw(self, CurrentSpaceType):
+        if self.callbacks.values().__len__() == 0:
+            return
+        if not self.returnTrueXAmountOfTimesPerSecondToCreateSolidFramerate(blender_imgui_global_framerate_cap):
+            self.imgui_backend.render(imgui.get_draw_data())
+            return
         context = bpy.context
         region = context.region
         io = imgui.get_io()
         io.display_size = region.width, region.height
         io.font_global_scale = context.preferences.view.ui_scale
+        # imgui.render()
+        # imgui.end_frame()
         imgui.new_frame()
-
         for cb, SpaceType in self.callbacks.values():
             if SpaceType == CurrentSpaceType:
                 cb(context)
-
+    
+        imguiFlags = (
+            imgui.WINDOW_NO_RESIZE
+            | imgui.WINDOW_NO_MOVE
+            | imgui.WINDOW_NO_COLLAPSE
+            | imgui.WINDOW_NO_TITLE_BAR
+            | imgui.WINDOW_ALWAYS_AUTO_RESIZE
+        )
+        imgui.begin("", closable=False, flags=imguiFlags)
```

Comment:
Could you comment why this empty window is needed?

Reply:
Hi @eliemichel , you can remove this if you feel its a little out of place.

 I added this window as I would sometimes see an error related to imgui.end_frame() never being called if I remember correctly (which I believe happened if I created a child class of BlenderImguiOverlay with an empty draw method in it, something that I would do in my code just to temporarily hide windows).

The alternative approach would be to "check" if imgui.begin was ever called in any of the registered draw methods and only call `imgui.end_frame()` (as well as `imgui.new_frame()` next time GlobalImgui.draw is called) if that is the case.

Reply:
Hi @eliemichel and thank you for the review! Its a rather silly solution for preventing an error caused by imgui.end_frame() being called before imgui.bein() resulting in an error and a hard crash if I remember correctly.

So its just an extra layer of safety in case imgui.being is never called in any draw function registered (say using your BlenderImguiOverlay).

I will let you choose yourself whether this is something you wanna keep or not, but this enables me to in my addon temporarily hide imgui windows by simply having an if statement that under certain conditions, simply never runs imgui.begin.